### ISSUE TITLE
Cholesky decomposition precision problem

### ISSUE THREAD

**Issue by markusr** [mathnet/mathnet-numerics#509]
I have the problem, that the Cholesky decomposition fails on some matrices. I think I narrowed it down to the fact that the values are small. 

See the following example:
```
        static void ShowCholeskyError()
        {
            var x1 = DenseMatrix.OfArray(new double[,] {{   0.0356381,  -0.000525888,  -4.85566E-05},
                                                       {-0.000525888,   1.49188E-05,  -3.29296E-05},
                                                       {-4.85566E-05,  -3.29296E-05,   0.000158206}});
            var chol1 = x1.Cholesky();
            Console.WriteLine("Cholesky1: {0}", chol1.Factor.ToString());

            var x2 = DenseMatrix.OfArray(new double[,] {{   0.035638057656433686,  -0.00052588845183103674,  -4.8556644746056749E-05},
                                                       {-0.00052588845183103674,   1.4918780214334419E-05,  -3.2929563148035619E-05},
                                                       {-4.8556644746056749E-05,  -3.2929563148035619E-05,   0.00015820643165769492}});
            //This will raise an exception
            var chol2 = x2.Cholesky(); 
            Console.WriteLine("Cholesky2: {0}", chol2.Factor.ToString());
        }
```

One workaround I found is:
```
            var x2_round = x2.Multiply(1e10).PointwiseRound().Divide(1e10);
            var chol2 = x2_round.Cholesky();
```

Using the same values of x2 other libraries (e.g. python numpy) don't fail.

**Comment by diluculo** [comment]
`x2` is not a positive definite matrix: `x2.Determinant( ) = -3.25E-26`. So, `L[2, 2]` can't be calculated.

**Comment by diluculo** [comment]
Regarding to the precision of Cholesky decomposition, I found the larger i, the larger the error of L[i, i]. It comes from roundoff error propagation at square of sqrt of x:

Let A and L be an input matrix and  its Cholesky factor, respectively.
`L11 = sqrt(A11)`
`L22 = sqrt(A22 - L21**2), where L21 = A12 / L11`
`L33 = sqrt(A33 - L31**2 - L32**2)`
and so on.

For more details, please see the [RosettaCode](https://rosettacode.org/wiki/Cholesky_decomposition), where implemented code avoids such an error.

**Comment by markusr** [comment]
Yes, you are correct. I tried to port some code from numpy to Mathnet and it showed an error. So I tried to reproduce the steps. I think the problem was with rounding of float64 which will lead to the ```Determinant < 0```.

**Comment by cdrnet** [comment]
It seems to me there is still something we could improve on our side?

**Comment by markusr** [comment]
I am not sure if it was really a bug in the first place. I just rechecked with python with the exact number as I posted.
```
import numpy as np

A = [[ 0.035638057656433686,  -0.00052588845183103674,  -4.8556644746056749E-05],
[-0.00052588845183103674,   1.4918780214334419E-05,  -3.2929563148035619E-05],
[-4.8556644746056749E-05,  -3.2929563148035619E-05,   0.00015820643165769492]]

L = np.linalg.cholesky(A)
```

This also raises the same error as Mathnet.
```
LinAlgError: Matrix is not positive definite
```

**Comment by tournierjc** [comment]
Hello,

I have solved this for Double case in my code by using an approximation:
1) Apply the eigen values decomposition (U = matrix of eigen vectors, A = vector of eigen values)
2) Transform A to a diagonal matrix D and apply a pointwise Max(0.0)
3) Apply QR decomposition to P = (U * D).Transpose
4) Return the R matrix which should be your approximated Cholesky decomposition (with a different sign)

I'm not familiar enough with the library yet (I started to use it yesterday), so I can't provide the fix myself.

`//Applu EVD decomposition (step 1)
Evd ev = M.Evd();

//Build D (step 2 & 3)
Matrix<double> D = Matrix<double>.Build.DenseDiagonal(evM.Count, evM.Count, i => Math.Sqrt(Math.Max(ev.EigenValues[i].Real, 0.0)));

//QR decomposition of (U*D).T (step 4)
Matrix<double> P = (ev.EigenVectors * D).Transpose();
return -1.0 * P.QR().R;`
